client/src/components/Focusedzone/Focusedzone.js

```diff
@@ -0,0 +1,84 @@
+/* global document */
+import React, { Component, PropTypes } from 'react';
+import ReactDOM from 'react-dom';
+
+/**
+ * Wrapper component that fires a handler (`onClickOut`) when a user clicks
+ * outside of the component.
+ *
+ *
+ * Example:
+ *
+ * ```jsx
+ *    class TestComponent extends React.Component {
+ *      ...
+ *
+ *      handleClickOut() {
+ *        alert('User clicked outside of TestComponent.');
+ *      }
+ *
+ *      render() {
+ *        return (
+ *          <Focusedzone onClickOut={this.handleClickOut}>
+ *            <div>... </div>
+ *          </Focusedzone>
+ *        );
+ *       }
+ *    < /Focusedzone>
+ *  ```
+ */
+
+class Focusedzone extends Component {
+
+  constructor(props) {
+    super(props);
+
+    this.wasClicked = false;
+    this.handleElementClick = this.handleElementClick.bind(this);
+    this.handleDocumentClick = this.handleDocumentClick.bind(this);
+  }
+
+  componentDidMount() {
+    const el = ReactDOM.findDOMNode(this);
+
+    // Click or touch event handlers on `el` and `document` will be both
+    // triggered when user clicked or touched inside the compoent's DOM. The
+    // handler on `el` is triggered before the one on `document` based on event
+    // bubbling.
+    ['click', 'touchstart'].forEach((eventName) => {
+      el.addEventListener(eventName, this.handleElementClick);
+      document.addEventListener(eventName, this.handleDocumentClick);
+    });
+  }
+
+  componentWillUnmount() {
+    const el = ReactDOM.findDOMNode(this);
+    el.removeEventHandler('click', this.handleElementClick);
+    el.removeEventHandler('click', this.handleDocumentClick);
+  }
+
+  handleElementClick() {
+    this.wasClicked = true;
+  }
+
+  handleDocumentClick() {
+    if (!this.wasClicked) {
+      this.props.onClickOut();
+    } else {
+      this.wasClicked = false;
+    }
+  }
+
+  render() {
+    return Array.isArray(this.props.children) ?
+      <div>{this.props.children}</div> :
+      React.Children.only(this.props.children);
+  }
+}
+
+Focusedzone.propTypes = {
+  children: PropTypes.object,
```

Comment:
I would suggest `PropTypes.any` for children, because children could be 0, 1 or many (class or function) components or even just a display string... and it's a headache to validate them all properly. `PropTypes.object` kind of masks the variable nature of `children`